src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
-			return;
-		}
-		/* Generate an IRQ if needed */
-		if (RampCtrl & 0x20) {
-			myGUS.RampIRQ|=irqmask;
-		}
-		/* Check for looping */
-		if (RampCtrl & 0x08) {
-			/* Bi-directional looping */
-			if (RampCtrl & 0x10) RampCtrl^=0x40;
-			RampVol = (RampCtrl & 0x40) ? (RampEnd-RampLeft) : (RampStart+RampLeft);
-		} else {
-			RampCtrl|=1;	//Stop the channel
-			RampVol = (RampCtrl & 0x40) ? RampStart : RampEnd;
-		}
-		UpdateVolumes();
+}
+
+void Gus::CheckIrq()
+{
+	if (irq_status && (mix_ctrl & 0x08))
+		PIC_ActivateIRQ(irq1);
+}
+
+bool Gus::CheckTimer(const size_t t)
+{
+	if (!timers[t].is_masked)
+		timers[t].has_expired = true;
+	if (timers[t].should_raise_irq) {
+		irq_status |= 0x4 << t;
+		CheckIrq();
 	}
+	return timers[t].is_counting_down;
+}
 
-	void generateSamples(Bit32s * stream,Bit32u len) {
-		//Disabled channel
-		if (RampCtrl & WaveCtrl & 3) return;
-
-		if (WaveCtrl & WCTRL_16BIT) {
-			for (int i = 0; i < (int)len; i++) {
-				// Get sample
-				Bit32s tmpsamp = GetSample16();
-				// Output stereo sample
-				stream[i << 1] += tmpsamp * VolLeft;
-				stream[(i << 1) + 1] += tmpsamp * VolRight;
-				WaveUpdate();
-				RampUpdate();
-			}
-		}
-		else {
-			for (int i = 0; i < (int)len; i++) {
-				// Get sample
-				Bit32s tmpsamp = GetSample8();
-				// Output stereo sample
-				stream[i << 1] += tmpsamp * VolLeft;
-				stream[(i << 1) + 1] += tmpsamp * VolRight;
-				WaveUpdate();
-				RampUpdate();
-			}
-		}
+void Gus::CheckVoiceIrq()
+{
+	irq_status &= 0x9f;
+	const Bitu totalmask = (voice_irq.vol_state | voice_irq.wave_state) &
+	                       active_voice_mask;
+	if (!totalmask)
+		return;
+	if (voice_irq.vol_state)
+		irq_status |= 0x40;
+	if (voice_irq.wave_state)
+		irq_status |= 0x20;
+	CheckIrq();
+	while (!(totalmask & 1ULL << voice_irq.status)) {
+		voice_irq.status++;
+		if (voice_irq.status >= active_voices)
+			voice_irq.status = 0;
 	}
-};
+}
 
-static GUSChannels *guschan[32];
-static GUSChannels *curchan;
-
-static void GUSReset(void) {
-	if((myGUS.gRegData & 0x1) == 0x1) {
-		// Reset
-		adlib_commandreg = 85;
-		myGUS.IRQStatus = 0;
-		myGUS.timers[0].raiseirq = false;
-		myGUS.timers[1].raiseirq = false;
-		myGUS.timers[0].reached = false;
-		myGUS.timers[1].reached = false;
-		myGUS.timers[0].running = false;
-		myGUS.timers[1].running = false;
-
-		myGUS.timers[0].value = 0xff;
-		myGUS.timers[1].value = 0xff;
-		myGUS.timers[0].delay = 0.080f;
-		myGUS.timers[1].delay = 0.320f;
-
-		myGUS.ChangeIRQDMA = false;
-		myGUS.mixControl = 0x0b;	// latches enabled by default LINEs disabled
-		// Stop all channels
-		int i;
-		for(i=0;i<32;i++) {
-			guschan[i]->RampVol=0;
-			guschan[i]->WriteWaveCtrl(0x1);
-			guschan[i]->WriteRampCtrl(0x1);
-			guschan[i]->WritePanPot(0x7);
-		}
-		myGUS.IRQChan = 0;
+uint32_t Gus::Dma8Addr()
+{
+	return static_cast<uint32_t>(dma_addr << 4);
+}
+
+uint32_t Gus::Dma16Addr()
+{
+	const auto lower = dma_addr & 0x1fff;
+	const auto upper = dma_addr & 0xc000;
```

Comment:
Suggestion; how about:

	const auto lower = dma_addr & 0b0001'1111'1111'1111;
	const auto upper = dma_addr & 0b1100'0000'0000'0000;

Also, are you sure one bit should not be covered by lower nor upper? I would expect these masks to be (`0x3fff` `0xc000`) or (`0x1fff` `0xe000`).

Reply:
I'm only using the existing values as-is; I'm not a bit-shifting guru, so have to defer to you here if we want to try to improve it (I'm happy to run tests and stand-alone code snippet comparisons and such).

Reply:
You can trying changing - these things are extremely fragile and audio will go way off the rails if things are off even slightly :-)

Reply:
Tried that, as well as:
``` c++
const auto lower = dma_addr & 0b0011'1111'1111'1111;
const auto upper = dma_addr & 0b1100'0000'0000'0000;
```
and
``` c++
const auto lower = dma_addr & 0b0001'1111'1111'1111;
const auto upper = dma_addr & 0b1110'0000'0000'0000;
```

But all regressed in quake generating white-noise.

Reply:
The bit range is "re-tightened" up after the _**and**_, but shifting by 1.

``` c++
uint32_t Gus::Dma16Addr()
{
	const auto lower = dma_addr & 0b0001'1111'1111'1111;
	const auto upper = dma_addr & 0b1100'0000'0000'0000;
	const auto combined = (lower << 1) | upper;
```

(and a similar one here, but with larger **and** values:
``` c++
	// Calculate offset of the 16-bit sample
	const auto lower = addr & 0b1100'0000'0000'0000'0000;
	const auto upper = addr & 0b0001'1111'1111'1111'1111;
	const size_t i = lower | (upper << 1);
```

I tried doing this:

``` c++
	const auto lower = addr & 0b1100'0000'0000'0000'0000;
	const auto upper = addr & 0b0011'1111'1111'1111'1111;
	const size_t i = lower | upper;
```

But get pure white-noise in 16-bit DMA reads.